gr-digital/lib/diff_stbc_encoder_cc_impl.cc

```diff
@@ -0,0 +1,119 @@
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#define _USE_MATH_DEFINES
+#include <cmath>
+#include <complex>
+
+#include <gnuradio/io_signature.h>
+#include "diff_stbc_encoder_cc_impl.h"
+
+namespace gr {
+  namespace digital {
+
+    diff_stbc_encoder_cc::sptr
+    diff_stbc_encoder_cc::make(float phase_offset)
+    {
+      return gnuradio::get_initial_sptr
+        (new diff_stbc_encoder_cc_impl(phase_offset));
+    }
+
+    /*
+     * The private constructor
+     */
+    diff_stbc_encoder_cc_impl::diff_stbc_encoder_cc_impl(float phase_offset)
+      : gr::sync_block("diff_stbc_encoder_cc",
+              gr::io_signature::make(1, 1, sizeof(gr_complex)),
+              gr::io_signature::make(2, 2, sizeof(gr_complex))),
+        d_basis_vecs(std::vector<gr_complex>(2, std::polar((float)M_SQRT1_2, phase_offset)))
+    {
+      d_mapping_coeffs = std::vector<gr_complex>(2, 0.0);
+      d_predecessor[0] = d_predecessor[1] = d_basis_vecs[0];
+      /* Set the number of input and output items to a multiple of 2,
+       * because the Alamouti algorithm processes sequences of 2 complex symbols.
+       */
+      set_output_multiple(2);
+    }
+
+    /*
+     * Our virtual destructor.
+     */
+    diff_stbc_encoder_cc_impl::~diff_stbc_encoder_cc_impl()
+    {
+    }
+
+    void
+    diff_stbc_encoder_cc_impl::map_symbols(const gr_complex *in) {
+      d_mapping_coeffs[0] =  in[0]*std::conj(d_basis_vecs[0]) + in[1]*std::conj(d_basis_vecs[1]);
+      d_mapping_coeffs[1] = -in[0]*          d_basis_vecs[1]  + in[1]*          d_basis_vecs[0];
+    }
+
+    void
+    diff_stbc_encoder_cc_impl::calculate_output(const gr_complex *in,
+                                        const gr_complex predecessor1,
+                                        const gr_complex predecessor2,
+                                        gr_complex *out1,
+                                        gr_complex *out2) {
+      // Transform input vector to new basis and calculate new coefficients.
+      map_symbols(in);
+      // Calculate the output of antenna 1.
+      out1[0] = d_mapping_coeffs[0]*predecessor1 - d_mapping_coeffs[1]*std::conj(predecessor2);
+      // Calculate the output of antenna 2.
+      out2[0] = d_mapping_coeffs[0]*predecessor2 + d_mapping_coeffs[1]*std::conj(predecessor1);
+      // Calculate the second element of the output sequence after the rules of Alamouti.
+      out1[1] = -std::conj(out2[0]);
+      out2[1] =  std::conj(out1[0]);
+    }
+
+    int
+    diff_stbc_encoder_cc_impl::work(int noutput_items,
+        gr_vector_const_void_star &input_items,
+        gr_vector_void_star &output_items)
+    {
+      const gr_complex *in = (const gr_complex *) input_items[0];
+      gr_complex *out1 = (gr_complex *) output_items[0];
+      gr_complex *out2 = (gr_complex *) output_items[1];
+
+      // Handle first sequence manually, because of a special predecessor assignment.
+      calculate_output(in, d_predecessor[0], d_predecessor[1], out1, out2);
```

Comment:
In the decoder, you handle blocks of input STBCs instead of single ones. Wouldn't this also simplify this work function? I'm again thinking of a function that receives the respective pointers and a length that is a multiple of 2 and then encodes symbols in a while() loop.